SPHNICS+ introduces some clang-tidy warnings

I fixed all of them on master except

```
Checked src/lib/pubkey/sphincsplus/sphincsplus_common/sp_treehash.cpp
/home/runner/work/botan/botan/src/lib/pubkey/sphincsplus/sphincsplus_common/sp_treehash.cpp:27:53: warning: the parameter 'out_auth_path' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
              StrongSpan<SphincsAuthenticationPath> out_auth_path,
                                                    ^
              const                                &
```

where I'm not sure if `clang-tidy`s suggestion is really correct. @reneme can you take a look at this?

--- a/src/lib/pubkey/sphincsplus/sphincsplus_common/sp_xmss.cpp
+++ b/src/lib/pubkey/sphincsplus/sphincsplus_common/sp_xmss.cpp
@@ -43,7 +43,7 @@ SphincsTreeNode xmss_sign_and_pkgen(StrongSpan<SphincsXmssSignature> out_sig,
 
    pk_addr.set_type(Sphincs_Address_Type::WotsPublicKeyCompression);
 
-   GenerateLeafFunction xmss_gen_leaf = [&](const StrongSpan<SphincsTreeNode>& out_root, TreeNodeIndex address_index) {
+   GenerateLeafFunction xmss_gen_leaf = [&](StrongSpan<SphincsTreeNode> out_root, TreeNodeIndex address_index) {
       wots_sign_and_pkgen(
          wots_bytes_s, out_root, secret_seed, address_index, idx_leaf, steps, leaf_addr, pk_addr, params, hashes);
    };

--- a/src/lib/utils/strong_type.h
+++ b/src/lib/utils/strong_type.h
@@ -528,13 +528,21 @@ class StrongSpan {
 
       StrongSpan(T& strong) : m_span(strong) {}
 
-      // allows implicit conversion from `StrongSpan<T>` to `StrongSpan<const T>`
-      // -> if T is a const type, the compiler will generate a copy-constructor
-      //    allowing copies from `StrongSpan<T>` and `StrongSpan<const T>`
-      // -> if T is a non-const type, no additional copy-constructor will be
-      //    generated. Instead this explicitly defines it. As a result, implicit
-      //    cast from a `StrongSpan<const T>` to `StrongSpan<T>` is prohibited.
-      StrongSpan(const StrongSpan<std::remove_const_t<T>>& other) : m_span(other.get()) {}
+      // Allows implicit conversion from `StrongSpan<T>` to `StrongSpan<const T>`.
+      // Note that this is not bi-directional. Conversion from `StrongSpan<const T>`
+      // to `StrongSpan<T>` is not allowed.
+      //
+      // TODO: Technically, we should be able to phrase this with a `requires std::is_const_v<T>`
+      //       instead of the `std::enable_if` constructions. clang-tidy (14 or 15) doesn't seem
+      //       to pick up on that (yet?). As a result, for a non-const T it assumes this to be
+      //       a declaration of an ordinary copy constructor. The existance of a copy constructor
+      //       is interpreted as "not cheap to copy", setting off the `performance-unnecessary-value-param` check.
+      //       See also: https://github.com/randombit/botan/issues/3591
+      template <concepts::contiguous_strong_type T2,
+                typename = std::enable_if_t<std::is_same_v<T2, std::remove_const_t<T>>>>
+      StrongSpan(const StrongSpan<T2>& other) : m_span(other.get()) {}
+
+      StrongSpan(const StrongSpan& other) = default;
 
       ~StrongSpan() = default;
 

--- a/src/lib/utils/strong_type.h
+++ b/src/lib/utils/strong_type.h
@@ -543,6 +543,11 @@ class StrongSpan {
        */
       underlying_span get() const { return m_span; }
 
+      /**
+       * @returns the underlying std::span without any type constraints
+       */
+      underlying_span get() { return m_span; }
+
       decltype(auto) data() noexcept(noexcept(this->m_span.data())) { return this->m_span.data(); }
 
       decltype(auto) data() const noexcept(noexcept(this->m_span.data())) { return this->m_span.data(); }

